File: 2024/Day_3/Solution.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Solution {
    private static final String FILE_PATH = "data.txt";
    public static void main(String[] args) {
        Solution solution = new Solution();
        String content = solution.readFileContent();
        
        if (content != null) {
            long partOneSum = solution.calculatePartOne(content);
            long partTwoSum = solution.calculatePartTwo(content);
            
            System.out.printf("Total sum for part 1: %d%n", partOneSum);
            System.out.printf("Total sum for part 2: %d%n", partTwoSum);
        }
    }

    private String readFileContent() {
        try {
            Path filePath = Paths.get(System.getProperty("user.dir"), FILE_PATH);
            return Files.readString(filePath);
        } catch (IOException e) {
            System.err.println("Error reading file: " + e.getMessage());
            return null;
        }
    }

    private long calculatePartOne(String content) {
        Pattern pattern = Pattern.compile("mul\\((\\d+),(\\d+)\\)");
        Matcher matcher = pattern.matcher(content);
        long total = 0;

        while (matcher.find()) {
            int num1 = Integer.parseInt(matcher.group(1));
            int num2 = Integer.parseInt(matcher.group(2));
            total += (long) num1 * num2;
        }

        return total;
    }

    private long calculatePartTwo(String content) {
        Pattern pattern = Pattern.compile("(?:mul\\((\\d+),(\\d+)\\)|do\\(\\)|don't\\(\\))");
        Matcher matcher = pattern.matcher(content);
        boolean enabled = true;
        long total = 0;

        while (matcher.find()) {
            String operation = matcher.group(0);
            
            if ("do()".equals(operation)) {
                enabled = true;
            } else if ("don't()".equals(operation)) {
                enabled = false;
            } else if (enabled && operation.startsWith("mul")) {
                int num1 = Integer.parseInt(matcher.group(1));
                int num2 = Integer.parseInt(matcher.group(2));
                total += (long) num1 * num2;
            }
        }

        return total;
    }
}

File: 2024/Day_3/Solution.ts
import * as fs from 'fs';
import * as path from 'path';

function readFileContent(filePath: string): string {
    try {
        return fs.readFileSync(path.join(process.cwd(), filePath), 'utf-8');
    } catch (error) {
        console.error('Error reading file:', error);
        return '';
    }
}

function calculatePartOne(content: string): number {
    const pattern = /mul\((\d+),(\d+)\)/g;
    let total = 0;
    let match: RegExpExecArray | null;
    
    while ((match = pattern.exec(content)) !== null) {
        const num1 = parseInt(match[1]);
        const num2 = parseInt(match[2]);
        total += num1 * num2;
    }
    
    return total;
}

function calculatePartTwo(content: string): number {
    const pattern = /(?:mul\((\d+),(\d+)\)|do\(\)|don't\(\))/g;
    let enabled = true;
    let total = 0;
    let match: RegExpExecArray | null;
    
    while ((match = pattern.exec(content)) !== null) {
        const operation = match[0];
        
        if (operation === "do()") {
            enabled = true;
        } else if (operation === "don't()") {
            enabled = false;
        } else if (enabled && operation.startsWith("mul")) {
            const num1 = parseInt(match[1]);
            const num2 = parseInt(match[2]);
            total += num1 * num2;
        }
    }
    
    return total;
}

function main(): void {
    const filePath = '2024/Day_3/data.txt';
    const content = readFileContent(filePath);
    
    if (content) {
        const partOneSum = calculatePartOne(content);
        const partTwoSum = calculatePartTwo(content);
        
        console.log(`Total sum for part 1: ${partOneSum}`);
        console.log(`Total sum for part 2: ${partTwoSum}`);
    }
}

main();